src/multipart/multipart.rs

```diff
@@ -0,0 +1,469 @@
+            Some(Cow::from(filename.to_string_lossy().into_owned()))
+        });
+        let file_length = std::fs::metadata(path.as_ref())
+            .ok()
+            .and_then(|metadata| Some(metadata.len() as u64));
+        Ok(MultipartField {
+            name: name.into(),
+            value: Box::new(std::fs::File::open(path.as_ref())?),
+            value_length: file_length,
+            mime: Some(::hyper::mime::APPLICATION_OCTET_STREAM),
+            file_name: file_name,
+        })
+    }
+    /// Sets the mime, builder style.
+    ///
+    /// ```
+    /// use reqwest::mime;
+    /// reqwest::MultipartField::param("key", "value").mime(Some(mime::IMAGE_BMP));
+    /// ```
+    ///
+    pub fn mime(mut self, mime: Option<Mime>) -> MultipartField {
+        self.mime = mime;
+        self
+    }
+    /// Sets the filename, builder style.
+    ///
+    /// ```
+    /// reqwest::MultipartField::param("key", "value").file_name(Some("filename"));
+    /// ```
+    ///
+    /// ```
+    /// let file_name = "file_name".to_string();
+    /// reqwest::MultipartField::param("key", "value").file_name(Some(file_name));
+    /// ```
+    ///
+    pub fn file_name<T: Into<Cow<'static, str>>>(mut self, filename: Option<T>) -> MultipartField {
+        self.file_name = filename.and_then(|filename| Some(filename.into()));
+        self
+    }
+    fn header(&self) -> String {
+        fn percent_encode(input: &str) -> String {
+            let mut result = String::with_capacity(input.len());
+            for character in input.chars() {
+                match character {
+                    // check for illegal ascii
+                    '%' => result.push_str("%25"),
+                    '"' => result.push_str("%22"),
+                    '\'' => result.push_str("%27"),
+                    '\r' => result.push_str("%0D"),
+                    '\n' => result.push_str("%0A"),
+                    // check for legal ascii
+                    other if other.is_ascii() => result.push(other),
+                    // must be non-ascii utf8
+                    other => {
+                        let mut buffer = [0; 4];
+                        for byte in other.encode_utf8(&mut buffer).bytes() {
+                            result.push_str(&format!("%{:X}", byte));
+                        }
+                    }
+                }
+            }
+            result
+        }
+
+        fn format_parameter(name: &str, value: &str) -> String {
+            let legal_value = percent_encode(value);
+            if value.len() == legal_value.len() {
+                // nothing has been percent encoded
+                format!("{}=\"{}\"", name, value)
+            } else {
+                // something has been percent encoded
+                format!("{}*=utf-8''{}", name, legal_value)
+            }
+        }
+
+        format!(
+            // TODO: I would use hyper's ContentDisposition header here, but it doesnt seem to have
```

Comment:
Hm, looking at the `ContentDisposition` header in hyper, it could use some help. It'd be a breaking change for hyper, so can't be changed for this pull request, unfortunately.